src/game/f3dex3.c

```diff
@@ -0,0 +1,388 @@
+    LinkedLight *light = alloc_display_list(sizeof(LinkedLight));
+    LinkedLight *nextLight = sLinkedLightHead;
+    light->pos[0] = x;
+    light->pos[1] = y;
+    light->pos[2] = z;
+
+    light->kc = kc;
+    light->kq = kq;
+    light->kl = kl;
+
+    light->col[0] = r;
+    light->col[1] = g;
+    light->col[2] = b;
+
+    light->size = size;
+
+    light->always = always;
+    light->next = NULL;
+
+    if (sLinkedLightHead == NULL) {
+        sLinkedLightHead = light;
+    } else {
+        while (nextLight->next != NULL) {
+            nextLight = nextLight->next;
+        }
+        nextLight->next = light;
+    }
+}
+
+/**
+  * Create a standard directional light.
+  *
+  * NOTE: These are given priority over point lights. You probably want at most 1-2 of these in a scene.
+  *
+  * @r      Red color component.
+  * @b      Blue color component.
+  * @g      Green color component.
+  *
+  * @x      X direction vector.
+  * @y      Y direction vector.
+  * @z      Z direction vector.
+  *
+  * @size   Defines the light's size for when using F3DEX3 specular shading.
+  */
+
+void add_directional_light(u8 r, u8 g, u8 b, s8 x, s8 y, s8 z, u8 size) {
+    LinkedDirectionalLight *light = alloc_display_list(sizeof(LinkedDirectionalLight));
+    LinkedDirectionalLight *nextLight = sLinkedDirectionalLightHead;
+
+    light->col[0] = r;
+    light->col[1] = g;
+    light->col[2] = b;
+
+    light->dir[0] = x;
+    light->dir[1] = y;
+    light->dir[2] = z;
+
+    light->size = size;
+    light->next = NULL;
+
+    if (sLinkedDirectionalLightHead == NULL) {
+        sLinkedDirectionalLightHead = light;
+    } else {
+        while (nextLight->next != NULL) {
+            nextLight = nextLight->next;
+        }
+        nextLight->next = light;
+    }
+}
+
+/**
+  * Create a global ambient light.
+  * NOTE: In the F3DEX3 microcode, ambient lights do not contribute towards the maximum amount of lights you can have.
+  * NOTE: If you do not set an ambient light and add lights to your scene, the lighting engine will add one with all colors set to 0.
+  *
+  * @r    Red color component.
+  * @g    Green color component.
+  * @b    Blue color component.
+  */
+
+void set_ambient_light(u8 r, u8 g, u8 b) {
+    if (sGlobalAmbientLight == NULL) {
+        sGlobalAmbientLight = alloc_display_list(sizeof(Ambient));
+    }
+
+    sGlobalAmbientLight->l.col[0] = r;
+    sGlobalAmbientLight->l.col[1] = g;
+    sGlobalAmbientLight->l.col[2] = b;
+
+    *(u32*)&sGlobalAmbientLight->l.colc[0] = *(u32*)&sGlobalAmbientLight->l.col[0];
+}
+
+static f32 calculate_distance_from_mario(s16 x, s16 y, s16 z) {
+    if (gMarioState->marioObj == NULL) { // No distance calc on Mario-less levels
+        return 0.0f;
+    }
+
+    f32 distX = ABS(gMarioState->pos[0] - (f32)x);
+    f32 distY = ABS(gMarioState->pos[1] - (f32)y);
+    f32 distZ = ABS(gMarioState->pos[2] - (f32)z);
+
+    return sqr(distX + distY + distZ);
+}
```

Comment:
Does it make more or less sense to use Mario state instead of Mario object here?